contracts/loot/adventurer/Adventurer.cairo

```diff
@@ -437,6 +438,89 @@ func increase_xp{
     return (TRUE,);
 }
 
+// @notice Allow purchasing health
+// @param adventurer_token_id: Id of adventurer
+// @return success: Value indicating success
+@external
+func allowPurchasingHealth{syscall_ptr: felt*, pedersen_ptr: HashBuiltin*, range_check_ptr, bitwise_ptr: BitwiseBuiltin*}(
+    adventurer_token_id: Uint256
+) {
+    Module.only_approved();
+
+    _set_purchasing_health(adventurer_token_id, TRUE);
+    return ();
+}
+
+
+// @notice Upgrade stat of adventurer
+// @param adventurer_token_id: Id of adventurer
+// @param amount: Amount of xp to increase
+// @return success: Value indicating success
+@external
+func upgradeStat{
+    syscall_ptr: felt*, pedersen_ptr: HashBuiltin*, range_check_ptr, bitwise_ptr: BitwiseBuiltin*
+}(adventurer_token_id: Uint256, stat: felt) -> (success: felt) {
+    alloc_locals;
+    // only adventurer owner can upgrade stat
+    ERC721.assert_only_token_owner(adventurer_token_id);
+
+    // unpack adventurer
+    let (unpacked_adventurer) = get_adventurer_by_id(adventurer_token_id);
+    let (adventurer_static_, adventurer_dynamic_) = AdventurerLib.split_data(unpacked_adventurer);
+    with_attr error_message("Adventurer: Adventurer must be upgradable") {
+        assert adventurer_dynamic_.Upgrading = TRUE;
+    }
+
+    // check stat is upgradeable
+    assert_in_range(stat, AdventurerSlotIds.Strength, AdventurerSlotIds.Luck);
+
+    // upgrade stat
+    let (updated_adventurer) = AdventurerLib.update_statistics(stat, adventurer_dynamic_);
+
+    // reset upgradability
+    _set_upgradable(adventurer_token_id, FALSE);
+
+    let (packed_new_adventurer: PackedAdventurerState) = AdventurerLib.pack(updated_adventurer);
+    adventurer_dynamic.write(adventurer_token_id, packed_new_adventurer);
+
+    return (TRUE,);
+}
+
+// @notice Purchase health for gold
+// @param adventurer_token_id: Id of adventurer
+// @return success: Value indicating success
+@external
+func purchaseHealth{
+    syscall_ptr: felt*, pedersen_ptr: HashBuiltin*, range_check_ptr, bitwise_ptr: BitwiseBuiltin*
+}( adventurer_token_id: Uint256) -> (success: felt) {
```

Comment:
we should make this dynamic, so users can purchase in increments of 10. So if you have 10 health you can purchase 9 potions in one transaction